### DISCUSSION TITLE
isDirty with empty number input (with valueAsNumber)

### DISCUSSION THREAD

**Discussion by tidusia** [react-hook-form/react-hook-form#5875] [Q&A] [answered]
Hi ! 

First, thanks for this great peace of software ! 🙏 

I struggle to make the `isDirty` state coherent with my actual form state.

I want : 
- an input number, that returns a number (with `valueAsNumber`)
- that can have an initial value, but if not, should leave the input empty (no `0`)
- if there is no default value, I should be able to fill the input, then empty it, and the `isDirty` state should returns back to `false`

I let an example of my problem here : https://codesandbox.io/s/peaceful-architecture-glpr9?file=/src/App.js

As you can see, if you just click and leave the input, the `isDirty` is wrongly set to `true`.

I know that this is because `valueAsNumber` returns back `NaN` which can never be equal to itself.

But is there any solution out there to make it work the way I wish ?

Thanks !

**Comment by bluebill1049** [comment]
<img width="1004" alt="Screen Shot 2021-07-09 at 11 54 51 am" src="https://user-images.githubusercontent.com/10513364/125011800-78c36480-e0ac-11eb-8439-656d6cd9af71.png">

this is most likely due to your defaultValue is `undefined`.

<img width="919" alt="Screen Shot 2021-07-09 at 11 55 37 am" src="https://user-images.githubusercontent.com/10513364/125011867-94c70600-e0ac-11eb-850a-441e9b500e03.png">

**Reply by tidusia** [reply]
@bluebill1049 thanks for you quick answer.

But the fact is that, if I give it a `null` value instead of `undefined`, then the input is filled with a `0`.

And my feature specification says that the input should be empty by default.

Any clue ?

EDIT: I changed the codesandbox example to reflects that.

**Reply by tidusia** [reply]
@bluebill1049 any clue to help me with this ?

**Comment by thanh-nguyen-95** [comment, marked as answer]
A working CSB: https://codesandbox.io/s/zen-leavitt-eseg0

You can use `setValueAs` option to manually parse the value. This way you can control the value more explicitly than using `valueAsNumber`.

I choose the default value when no input is given as `null`. You can change it to whatever suits your needs.

**Reply by tidusia** [reply]
Oh real thanks @thanh-nguyen-95 that completely do the work 🙏
